src/Dom/Document.php

```diff
@@ -480,21 +495,13 @@ private function reset()
      */
     public function loadHTML($source, $options = [])
     {
-        $source  = $this->normalizeDocumentStructure($source);
-        $success = $this->loadHTMLFragment($source, $options);
-
-        if ($success) {
-            $this->insertMissingCharset();
-
-            // Do some further clean-up.
-            $this->deduplicateTag(Tag::HEAD);
-            $this->deduplicateTag(Tag::BODY);
-            $this->moveInvalidHeadNodesToBody();
-            $this->movePostBodyNodesToBody();
-            $this->convertHeadProfileToLink();
-        }
+        $this->registerMiddlewareClass(NormalizeDocumentStructure::class, 9);
+        $this->registerMiddlewareClass(InsertMissingCharset::class, 11);
+        $this->registerMiddlewareClass(DeDuplicateTag::class, 11);
+        $this->registerMiddlewareClass(MoveInvalidNodesToBody::class, 11);
+        $this->registerMiddlewareClass(HeadProfileToLink::class, 11);
```

Comment:
Why are these not part of the central middleware registration?

Reply:
These middlewares will only used when we use the `loadHtml` method. The middlewares registered in the constructor will be used for both the `loadHtml` and `loadHTMLFragment` methods.

Reply:
Ah, I see. I would keep these as part of the `Document` for now and not turn them into middleware, because:
- they are not fixing a bug in the DOM implementation, but rather enforcing assumptions we make about the ideal AMP HTML structure;
- they are a one-off change to the HTML and do not need restoring afterward;
- they are less of a "fix a bug", and more of a "make Dom\Document opinionated".